python/lsst/ip/diffim/detectAndMeasure.py

```diff
@@ -283,20 +294,29 @@ def run(self, science, matchedTemplate, difference,
             ``diaSources``  : `lsst.afw.table.SourceCatalog`
                 The catalog of detected sources.
         """
+        if idFactory is None:
+            idFactory = lsst.meas.base.IdGenerator().make_table_id_factory()
```

Comment:
I don't know if `IdGenerator` is immutable: there's always a danger assigning a mutable object as the kwarg default because every use of the default would use the same object reference.

Reply:
Might be worth considering making it not have a default - the fallback one should never be used in production, as it doesn't encode the data ID at all.

Reply:
The problem with that is that it makes it harder to call the task outside of a production environment (e.g. in a notebook). We've had a convention that `run` is easily callable with just the data to be processed in most Tasks, and I'd like to maintain that.

Reply:
I agree with Jim that this is dangerous to leave with a default value. Perhaps to support running in a notebook we could have a config such as `useFakeIdGenerator`, and set `idFactory` as you have here if that is set and raise a helpful warning if it's not and `idFactory` is None?

Reply:
`if [idFactory|idGenerator] is None` at the top of `run` is the convention we have in most tasks that use one (one's that I found in a quick search: characterizeImage, calibrate, calibrateImage, processCcdWithFakes, simpleAssociation, forcedMeasurement). I don't see the problem: this is not a parameter that a user should be expected to configure and set when running such a task in a notebook, and it is passed in from `runQuantum`, so will be set when running pipetask.

Reply:
> If we want to make the task run methods for our detection/measurement tasks more cumbersome to call outside of a running pipeline, we should have a broader conversation about that.

FWIW, I kind of do (or at least I want to promote the idea that running tasks from notebooks is much less important than running them in production, and see how often various members of the team _actually_ ever call `run` directly); I'm inclined to include that in the "historical traps we've fallen into" discussion we've been mulling.

I don't feel that strongly about this case, but I'd rather leave it defaulted than introduce another config field for it.